src/forks.rs

```diff
@@ -0,0 +1,230 @@
+                println!("new_trunk id {}", new_trunk.1.fork_id());
+                return Err(BankError::CheckpointNotFinalized);
+            }
+            if !old_trunk.1.finalized() {
+                println!("old id {}", old_trunk.1.fork_id());
+                return Err(BankError::CheckpointNotFinalized);
+            }
+            //stupid sanity checks
+            assert_eq!(new_trunk.1.fork_id(), new_trunk.0);
+            assert_eq!(old_trunk.1.fork_id(), old_trunk.0);
+            (old_trunk.1.clone(), new_trunk.1.clone(), new_trunk.0)
+        };
+        let idag = self.checkpoints.invert();
+        let new_checkpoints = self.checkpoints.prune(new_trunk_id, &idag);
+        let old_trunk_id = old_trunk.fork_id();
+        self.checkpoints = new_checkpoints;
+        self.trunk_fork = new_trunk_id;
+        self.live_fork = leaf;
+        // old should have been pruned
+        assert!(self.checkpoints.load(old_trunk_id).is_none());
+        // new_trunk id should be in the new tree
+        assert!(!self.checkpoints.load(new_trunk_id).is_none());
+
+        // swap in the old instance under the new_trunk id
+        // this should be the last external ref to `new_trunk`
+        self.checkpoints
+            .insert(new_trunk_id, old_trunk.clone(), old_trunk_id);
+
+        // merge all the new changes into the old instance under the new id
+        // this should consume `new`
+        // new should have no other references
+        let new_trunk: BankCheckpoint = Arc::try_unwrap(new_trunk).unwrap();
+        old_trunk.merge_into_trunk(new_trunk);
+        assert_eq!(old_trunk.fork_id(), new_trunk_id);
+        Ok(new_trunk_id)
+    }
+
+    /// Initialize the first trunk
+    pub fn init_trunk_fork(&mut self, checkpoint: BankCheckpoint) {
+        assert!(self.checkpoints.is_empty());
+        self.live_fork = checkpoint.fork_id();
+        self.trunk_fork = checkpoint.fork_id();
+        //TODO: using u64::MAX as the impossible checkpoint
+        //this should be a None instead
+        self.checkpoints
+            .store(self.live_fork, Arc::new(checkpoint), std::u64::MAX);
+    }
+
+    pub fn is_active_fork(&self, fork: u64) -> bool {
+        if let Some(state) = self.checkpoints.load(fork) {
+            !state.0.finalized() && self.live_fork == fork
+        } else {
+            false
+        }
+    }
+    /// Initalize the `current` fork that is a direct decendant of the `base` fork.
+    pub fn init_fork(&mut self, current: u64, last_id: &Hash, base: u64) -> Result<()> {
+        if let Some(state) = self.checkpoints.load(base) {
+            if !state.0.finalized() {
+                return Err(BankError::CheckpointNotFinalized);
+            }
+            let new = state.0.fork(current, last_id);
+            self.checkpoints.store(current, Arc::new(new), base);
+            self.live_fork = current;
+            Ok(())
+        } else {
+            return Err(BankError::UnknownFork);
```

Comment:
Typical style is to put this branch first and then not intent the success case.

Reply:
@garious there is no clean `!let Some(...) =`